pkg/controller/chaosscheduler/createEngineForRepeat.go

```diff
@@ -128,50 +135,44 @@ func (schedulerReconcile *reconcileScheduler) createNewEngine(cs *chaosTypes.Sch
 	}
 	cs.Instance.Status.Schedule.StartTime = startTime
 
-	if errUpdate := schedulerReconcile.r.client.Update(context.TODO(), cs.Instance); errUpdate != nil {
-		return reconcile.Result{}, errUpdate
+	if err := schedulerReconcile.r.client.Update(context.TODO(), cs.Instance); err != nil {
+		return reconcile.Result{}, err
 	}
 
 	return reconcile.Result{}, nil
 }
 
-func getRecentUnmetScheduleTime(cs *chaosTypes.SchedulerInfo, cronString string) (*time.Time, error) {
+func (schedulerReconcile *reconcileScheduler) getRecentUnmetScheduleTime(cs *chaosTypes.SchedulerInfo, cronString string) (*time.Time, time.Duration, error) {
 
 	now := time.Now()
 	cronSchedule, err := cron.ParseStandard(cronString)
 	if err != nil {
-		return nil, fmt.Errorf("unparseable schedule: %s : %s", cronString, err)
+		return nil, 0, fmt.Errorf("unparseable schedule: %s : %s", cronString, err)
 	}
 
 	timeRange := cs.Instance.Spec.Schedule.Repeat.TimeRange
-
+	var nextTime *time.Time
 	var earliestTime time.Time
+
 	if cs.Instance.Status.LastScheduleTime != nil {
+		// handles all the schedules except first schedule
 		earliestTime = cs.Instance.Status.LastScheduleTime.Time
-	} else if timeRange != nil && timeRange.StartTime != nil {
-		// If none found, then this is either a recently created schedule,
-		// or the active/completed info was somehow lost (it may need to be recreated),
-		// or that we have started a engine, but have not noticed it yet (distributed
-		//	systems can have arbitrary delays).
-		// Since we are checking the startTime with currentTime. It is
-		// possible that we might miss a schedule at the TIME of CREATION
-		// by just a few seconds because of the slight delay in reconciliation process.
-		earliestTime = timeRange.StartTime.Time.Add(time.Minute * -1)
+	} else if timeRange != nil && timeRange.StartTime != nil && !cs.Instance.GetCreationTimestamp().Time.After(timeRange.StartTime.Time) {
+		// handles the first schedule
+		earliestTime = timeRange.StartTime.Time
 	} else {
+		// handles the first schedule
 		earliestTime = cs.Instance.GetCreationTimestamp().Time
 	}
-
-	if earliestTime.After(now) {
-		return nil, nil
+	t := cronSchedule.Next(earliestTime)
+	if !t.After(now) {
+		nextTime = &now
```

Comment:
What if becaus of some mishappening LastScheduledTime is not updated.

Lets say we have engines to be scheduled at 1, 3, 5, 7, 9 minutes and when we reach 5th min because of some update mishappening lastScheduledTime is still 1st min. Then from 5th to 9th min each and every schedule will fail.
That should not be the ideal case. Thats why there was a loop previously.

Reply:
Yes, we can have a loop but because of that loop, the scheduled time changed a little bit. Will check and handle this case.

Reply:
What is the problem with the loop ?? I didn't get it.

Reply:
The loop finds the next time wrt now() time. As now(() time is not constant so it may miss the schedule